examples/imitation_learning_openbot/generate_dataset.py

```diff
@@ -217,16 +180,16 @@
                 # the new goal position. Doing so requires a recomputation of the compass observation, since the latter is goal dependant. Therefore, rather
                 # than directly writing all the observations in a file iteration by iteration, we append these observations in a buffer, named "observation"
                 # to later process them once the episode is completed. 
-                control_data[i]     = obs["control_data"]                  # control_data: [ctrl_left, ctrl_right]
-                state_data[i]       = obs["state_data"]                    # state_data: [x, y, z, pitch, yaw, roll]
-                waypoint_data[i]    = policy_step_info["current_waypoint"] # current waypoint being tracked by the agent
-                time_data[i]        = time_stamp                           # current time stamp
-                frame_data[i]       = i                                    # current frame
+                control_data[i]     = action                                                                    # control_data: [ctrl_left, ctrl_right]
+                state_data[i]       = np.concatenate((obs["location"], np.deg2rad(obs["rotation"])), axis=None) # state_data: [x, y, z, pitch, yaw, roll]
+                waypoint_data[i]    = policy_step_info["current_waypoint"]                                      # current waypoint being tracked by the agent
+                time_data[i]        = time_stamp                                                                # current time stamp
+                frame_data[i]       = i                                                                         # current frame
 
             # debug
             if args.debug:
                 show_obs(
-                    obs, config.SIMULATION_CONTROLLER.OPENBOT_AGENT.OBSERVATION_COMPONENTS, config.SIMULATION_CONTROLLER.OPENBOT_AGENT.CAMERA.RENDER_PASSES)
+                    obs, config.SIMULATION_CONTROLLER.VEHICLE_AGENT.OBSERVATION_COMPONENTS, config.SIMULATION_CONTROLLER.VEHICLE_AGENT.CAMERA.RENDER_PASSES)
 
             # termination conditions
```

Comment:
It's fine for a policy to compute `waypoint_reached` because that is an implementation detail of the policy, and it is just used for debugging. But these policies should not be computing `goal_reached`, because that is an application-level responsibility.

```
            # update control action 
            action, policy_step_info = policy.step(obs)

            # send control action to the agent and collect observations
            obs, _, _, env_step_info = env.step(action={"set_duty_cycles": action})

            # compute goal reached condition
            # ...

            num_iterations = num_iterations + 1

            if args.debug:
                show_obs(obs, config.SIMULATION_CONTROLLER.VEHICLE_AGENT.OBSERVATION_COMPONENTS, config.SIMULATION_CONTROLLER.VEHICLE_AGENT.CAMERA.RENDER_PASSES)

            if not args.benchmark:
                obs_final_color = obs["camera.final_color"]
                assert len(obs_final_color.shape) == 3
                assert obs_final_color.shape[2] == 4
                obs_final_color = obs_final_color[:,:,[2,1,0,3]].copy() # note that spear.Env returns BGRA by default

                # save the collected rgb observations
                plt.imsave(os.path.realpath(os.path.join(image_dir, "%04d.jpg"%i)), obs_final_color)

                # During an episode, there is no guarantee that the agent reaches the predefined goal although its behavior is perfectly valid for training
                # purposes. In practice, it may for instance occur that the agent is not given enough time steps or control authority to move along the whole
                # trajectory. In this case, rather than considering the whole episode as a fail, one can consider the last position reached by the agent as
                # the new goal position. Doing so requires a recomputation of the compass observation, since the latter is goal dependant. Therefore, rather
                # than directly writing all the observations in a file iteration by iteration, we append these observations in a buffer, named "observation"
                # to later process them once the episode is completed. 
                control_data[i]     = action                                                                    # control_data: [ctrl_left, ctrl_right]
                state_data[i]       = np.concatenate((obs["location"], np.deg2rad(obs["rotation"])), axis=None) # state_data: [x, y, z, pitch, yaw, roll]
                waypoint_data[i]    = policy_step_info["current_waypoint"]                                      # current waypoint being tracked by the agent
                time_data[i]        = time_stamp                                                                # current time stamp
                frame_data[i]       = i                                                                         # current frame

            # termination conditions
            if env_step_info["task_step_info"]["hit_obstacle"]: 
                spear.log("    Collision detected, ending episode...")
                collect_data_for_episode = False
                break
            elif env_step_info["task_step_info"]["hit_goal"]: 
                spear.log("    Goal reached according env_step_info, ending episode...")
                break
            elif goal_reached: 
                spear.log("    Goal reached according to local distance calculation.")
                break
```

Reply:
Update this